### PR TITLE
JDK-8290137: riscv: small refactoring for add_memory_int32/64

### PR DESCRIPTION
**Pull request by RealFYang**
Currently, add_memory_int32/64 for riscv can only add a sign-extended 12-bit immediate to memory since they call addi/addiw assembler direcly. This constraint could be relaxed when the given memory address is in the expected form: base register plus a sign-extended 12-bit offset. In this case, we can emit code for load + add/sub + store sequence adding arbitrary immediate to memory with no more than two scratch registers (t0 and t1) available.

We could also refactor these two functions into four seperate functions: increment, incrementw, decrement and decrementw, so that it will be more clear in code logic at the call sites.

Testing: tier1 tested on riscv64-linux unmatched board.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290137](https://bugs.openjdk.org/browse/JDK-8290137): riscv: small refactoring for add_memory_int32/64

### Reviewers
 * [Yadong Wang](https://openjdk.org/census#yadongwang) (@yadongw - Author)
 * [Feilong Jiang](https://openjdk.org/census#fjiang) (@feilongjiang - Author)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9461/head:pull/9461` \
`$ git checkout pull/9461`

Update a local copy of the PR: \
`$ git checkout pull/9461` \
`$ git pull https://git.openjdk.org/jdk pull/9461/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9461`

View PR using the GUI difftool: \
`$ git pr show -t 9461`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9461.diff">https://git.openjdk.org/jdk/pull/9461.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 1]
```diff
@@ -2955,19 +2955,47 @@ Address MacroAssembler::add_memory_helper(const Address dst) {
...
+void MacroAssembler::increment(const Address dst, int64_t value) {
+  assert(((dst.getMode() == Address::base_plus_offset &&
+           is_offset_in_range(dst.offset(), 12)) || is_imm_in_range(value, 12, 0)),
+          "invalid value and address mode combination");
   Address adr = add_memory_helper(dst);
-  assert_different_registers(adr.base(), t0);
+  assert(!adr.uses(t0), "invalid dst for address increment");
   ld(t0, adr);
-  addi(t0, t0, imm);
+  add(t0, t0, value, t1);
```
Comment (yadongw) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2965]: It's not safe to clobber t1 sometimes. And I think it's better to limit the accepted value to 12 bits or less.

Comment (RealFYang) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2965]: In fact, we still need to trash t1 here when we have some unexpected memory address. Callers should be aware of this.

**Review by yadongw — APPROVED** [review]

**Comment by RealFYang** [comment]
Could we have a Reviewer please? Maybe @shipilev ?

File: src/hotspot/cpu/riscv/c1_LIRAssembler_arraycopy_riscv.cpp [hunk 2]
```diff
@@ -57,7 +57,7 @@ void LIR_Assembler::generic_arraycopy(Register src, Register src_pos, Register l
   __ mv(c_rarg4, j_rarg4);
 #ifndef PRODUCT
   if (PrintC1Statistics) {
-    __ add_memory_int32(ExternalAddress((address)&Runtime1::_generic_arraycopystub_cnt), 1);
+    __ incrementw(ExternalAddress((address)&Runtime1::_generic_arraycopystub_cnt));
```
Comment (shipilev) [inline src/hotspot/cpu/riscv/c1_LIRAssembler_arraycopy_riscv.cpp:60]: Is it really `incrementw`, though? These counter fields are `int`-s, are they still 32-bit on RISC-V? If so, shouldn't it be `incrementl`?

**Review by shipilev — APPROVED** [review]
Looks fine, with one minor question.

File: src/hotspot/cpu/riscv/c1_LIRAssembler_arraycopy_riscv.cpp (hunk 2 above)
Comment (RealFYang) [reply src/hotspot/cpu/riscv/c1_LIRAssembler_arraycopy_riscv.cpp:60]: @shipilev : Yes, the type of _generic_arraycopystub_cnt is int and it occupies 32-bit in memory. That's why we use incrementw here which increments a 32-bit memory operand. Note that incrementl works for 64-bit memory operand. Hope that explains. Thanks.

Comment (shipilev) [reply src/hotspot/cpu/riscv/c1_LIRAssembler_arraycopy_riscv.cpp:60]: Is this RISC-V specific postfix naming? On x86, there is `b`/`w`/`l`/`q` for 1/2/4/8-byte ops, respectively. This is just my curiosity, it does not block the integration.

Comment (RealFYang) [reply src/hotspot/cpu/riscv/c1_LIRAssembler_arraycopy_riscv.cpp:60]: Yes, default is for 8-byte ops and we use 'w' post-fix for 4-byte ops on RISC-V. And I think this is hangover from the aarch64 port [1]. 

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp#L1917

**Pull request closed without merging**